Documentation/api-design-guide.md

```diff
@@ -0,0 +1,461 @@
+
+// Actual definition below.
+USTRUCT(BlueprintType)
+struct CESIUMRUNTIME_API FCesiumMetadataPrimitive { ... }
+```
+
+- For backwards compatibility, sometimes you'll need to keep references to the deprecated classes or functions in C++ code. If this is the case, then be sure to wrap the relevant lines in `PRAGMA_DISABLE_DEPRECATION_WARNINGS` and `PRAGMA_ENABLE_DEPRECATION_WARNINGS`. This will reduce the spam in the Unreal logs that can otherwise occur during the build process.
+
+```c++
+struct LoadPrimitiveResult { 
+  // List of properties here...
+
+  PRAGMA_DISABLE_DEPRECATION_WARNINGS
+  // For backwards compatibility with CesiumEncodedMetadataComponent.
+  FCesiumMetadataPrimitive Metadata_DEPRECATED{};
+  PRAGMA_ENABLE_DEPRECATION_WARNINGS
+
+  // Other properties here...
+};
+```
+
+### Custom Versions
+
+Sometimes a property on a `UObject` is deprecated, or replaced by an alternative solution. However, it may have been set by the user with some value that should successfully translate to the new implementation. This can be resolved by creating "custom versions" of the plugin that are specially handled at load time.
+
+Custom versions boil down to an `enum` called `Versions` that captures the current state of the plugin. `Versions` is stored in `CesiumCustomVersion.h` and can be expanded in the future. When objects are loaded from serialization, their properties can be specially handled if saved under a previous version.
+
+We'll walk through an existing code example below using version `WebMapTileServiceProjectionAsEnum`. In this version, a boolean property on `UCesiumWebMapTileServiceRasterOverlay` was replaced with an enum.
+
+First, deprecate the old property by following the guide linked in [Deprecation Specifiers](#deprecation-specifiers). Use the [Core Redirects](#core-redirects) as needed to redirect properties to their `_DEPRECATED` counterparts.
+
+Then, create a new version in the `Versions` enum. The enum value should:
+  - Have a useful name that briefly summarizes the nature of the change.
+  - Have a comment with further explanation.
+  - Appear between the last custom version and `VersionPlusOne`.
+
+```cpp
+enum Versions {
+    // Other versions omitted.
+    // [...]
+
+    // Replaced the UseWebMercatorProjection property in
+    // CesiumWebMapTileServiceOverlay with the enum Projection property.
+    WebMapTileServiceProjectionAsEnum = 8,
+
+    VersionPlusOne,
+    LatestVersion = VersionPlusOne - 1
+}
+```
+
+After that, add the `virtual` `Serialize` function to the target class (if it does not yet exist).
+
+```cpp
+class UCesiumWebMapTileServiceRasterOverlay {
+  protected:
+    virtual void Serialize(FArchive& Ar) override;
+}
+```
+
+In `Serialize`, you can retrieve the version associated with the `UObject` and implement compatibility measures. Here, the deprecated value of `UseWebMercatorProjection` (which has been redirected by `[CoreRedirects]`) is used to set the new property as needed.
```

Comment:
At the time `Serialize` is called, properties that refer to other objects will not reliably be set, and attempting to set one risks it being overwritten later in the load process. So if any of that is necessary, the solution is to also override `PostLoad` and use `GetLinkerCustomVersion` to determine if upgrading is necessary. We have a few examples of this.